packages/grafana-data/src/events/EventBus.ts

```diff
@@ -90,3 +90,70 @@ export class EventBusSrv implements EventBus, LegacyEmitter {
     this.emitter.removeAllListeners();
   }
 }
+
+/**
+ * @alpha
+ *
+ * Wraps EventBus and adds a source path to help with identifying if a subscriber should react to the event or not.
+ */
+export class EventBusWithSource implements EventBus {
+  private _sourceFragment: string;
+  get source(): string[] {
+    if (this.eventBus instanceof EventBusWithSource) {
+      return this.eventBus.source.concat([this._sourceFragment]);
+    }
+    return [this._sourceFragment];
+  }
+
+  eventBus: EventBus;
+
+  constructor(eventBus: EventBus, sourceFragment: string) {
+    this.eventBus = eventBus;
+    this._sourceFragment = sourceFragment;
+  }
+
+  publish<T extends BusEvent>(event: T): void {
+    const payload = event.payload ?? { source: [] };
+    const decoratedEvent = {
+      ...event,
+      ...{ payload: { ...payload, ...{ source: [...[this._sourceFragment], ...(payload.source ?? [])] } } },
+    };
+    this.eventBus.publish(decoratedEvent);
+  }
+
+  subscribe<T extends BusEvent>(eventType: BusEventType<T>, handler: BusEventHandler<T>): Unsubscribable {
+    return this.eventBus.subscribe(eventType, handler);
+  }
+
+  getStream<T extends BusEvent>(eventType: BusEventType<T>): Observable<T> {
+    return this.eventBus.getStream(eventType);
+  }
+
+  removeAllListeners(): void {
+    this.eventBus.removeAllListeners();
+  }
+
+  /**
+   * Appends a source fragment id to the source
+   *
+   * @param sourceFragmentId source id to append to the eventbus
+   * @returns a new instance of EventBusWithSource with the new source fragment appended
+   */
+  appendSource(sourceFragmentId: string) {
+    return new EventBusWithSource(this, sourceFragmentId);
+  }
```

Comment:
do we have a use case for this? does not look like it's used

Reply:
Nope, right now I think we only need the reference to the panel-if. It only adds unnecessary complexity, I’ll go ahead and remove it.